source/adapters/cuda/usm.cpp

```diff
@@ -159,6 +159,12 @@ ur_result_t USMSharedAllocImpl(void **ResultPtr, ur_context_handle_t,
     ScopedContext Active(Device);
     UR_CHECK_ERROR(cuMemAllocManaged((CUdeviceptr *)ResultPtr, Size,
                                      CU_MEM_ATTACH_GLOBAL));
+    for (const auto &device : Context->getDevices()) {
```

Comment:
nit: I don't claim to know anything about the coding style used here, but we have `Context`, `Device`, `Size,` and `Err` in this function, and we're introducing lower-case `device` here. Is it not inconsistent, and isn't it confusing to have two uses of "device" in the same function?

Reply:
> Agreed. I think in general good to go for `hDevice` for the device arg and then `Dev` for the local device. But I don't think the device arg is actually needed? Maybe better to do:
> 
> ```
> ScopedContext Active(Context->getDevices()[0]);
> ```
> 
> And just remove the `Device` arg altogether

OK I'll switch the naming. But I don't think any device param can be removed. 

What is happening here is that a stream within CUcontext of the Device that was associated with a runtime (sycl) queue is allocating managed memory. Probably in many cases we could ignore this prescription provided by the programmer and just use the first device from the `Context->getDevices()[0]` and this wouldn't matter to a user. However I think it could do. 
Say for example an application from User A is on a node but sees all devices in the platform on that node (in perlmutter there are 4 gpus). If the Context corresponds to a sycl default context then all four gpus will be in the context. The user provides device 1 as the device whos allocated CUcontext they prescribe to allocate the managed memory via. User A only wants to use device 0. We mark accessed_by for all devices but this isn't used - not a big deal, it is cheap
User B is using device 0 for a different application on the same node. I imagine that if user B is saturating the gpu, it might be a bad idea for us to break the prescription from user A and use device 0 instead.
Probably it doesn't make a big difference in practice, but hopefully the above details there is a potential difference. If the interface used provides a specific device, then I think this should be used as the "command device"

Reply:
In that case we should also be calling `setAccessedBy` for the `hDevice` param, or asserting that the device is within the context

Reply:
OK I'll add an assert.

Reply:
I've just ended up calling the advise also via the commandDevice. This will lead to a duplication in normal circumstances, but I've checked you can advise twice without an error.

Reply:
Yeah that sounds good to me.